zerver/tests/test_auth_backends.py

```diff
@@ -2571,6 +2571,37 @@ def test_dev_ldap_fail_login(self) -> None:
 
             assert(user_profile is None)
 
+    @override_settings(AUTHENTICATION_BACKENDS=('zproject.backends.ZulipLDAPAuthBackend',))
+    def test_email_belongs_to_ldap(self) -> None:
+        realm = get_realm("zulip")
+        # Config A:
+        with self.settings(
+            AUTH_LDAP_USER_SEARCH = LDAPSearch("ou=users,dc=zulip,dc=com",
+                                               ldap.SCOPE_ONELEVEL, "(uid=%(user)s)"),
+            AUTH_LDAP_REVERSE_EMAIL_SEARCH = LDAPSearch("ou=users,dc=zulip,dc=com",
+                                                        ldap.SCOPE_ONELEVEL, "(uid=%(email)s)"),
+        ):
+            self.assertTrue(email_belongs_to_ldap(realm, "newuser_email_as_uid@zulip.com"))
+            self.assertFalse(email_belongs_to_ldap(realm, "nonexistent@zulip.com"))
+
+        # Config B:
+        with self.settings(
+            AUTH_LDAP_USER_SEARCH = LDAPSearch("ou=users,dc=zulip,dc=com",
+                                               ldap.SCOPE_ONELEVEL, "(uid=%(user)s)"),
+            LDAP_APPEND_DOMAIN='zulip.com'
+        ):
+            self.assertTrue(email_belongs_to_ldap(realm, self.example_email("hamlet")))
+            # If the email isn't in ldap, return False, even though the domain matches LDAP_APPEND_DOMAIN
+            self.assertFalse(email_belongs_to_ldap(realm, "nonexistent@zulip.com"))
+            # Test a non-matching domain too:
+            self.assertFalse(email_belongs_to_ldap(realm, "hamlet@example.com"))
+            self.assertFalse(email_belongs_to_ldap(realm, "nonexistent@example.com"))
```

Comment:
Hmm.  So, I'm not sure the the behavior change here is great.  A common use case for `LDAP_APPEND_DOMAIN` + `EmailAuthBackend` is that you're using `EmailAuthBackend` for guests/contractors who might be outside your email domain, while trying to manage everyone in your domain via LDAP.  

In that case, you very much don't want it to be possible to e.g. create an account within your domain that isn't managed by LDAP.

Reply:
Hmm, trying to keep this behavior in ``email_belongs_to_ldap`` made it no longer particularly adventagous to call ``django_to_ldap_username`` in the function, so I got rid of the commit.

But actually, we weren't protecting against creating a non-ldap account with email matching LDAP_APPEND_DOMAIN in the registration flow, so I added a commit to fix that.